File: tests/run_tests.cpp
// Test driver for the grade book program.
//
// Runs the grade book once per file in tests/inputs (in name order). The driver
// is linked with Menu, Semester and StudentRecord, and for each run it points cin
// at the input file and cout at screen.txt, so the menu reads the file as if it
// were typed. All runs share one working folder, so the Grades.dat saved by one
// run is loaded by the next, like running the program several times by hand.
// After each run, Grades.dat, Grades.out, Grades.trn and screen.txt are copied to
// test_runs/<run name>/, since the program rewrites them every run.
//
// Each run's Grades files are then compared with the known-good copies in
// tests/expected/<run name>/, and the run passes if they match. After checking a
// deliberate change by hand, run with --update to save the new output as expected.
//
// Every input file must end with Q: the menu does not stop at end of input, so a
// missing Q makes the driver hang.
//
// usage: run_tests [--update]
#include "Menu.h"
#include <algorithm>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

namespace fs = std::filesystem;

// PROJECT_ROOT is set by CMakeLists.txt to the Program4 folder
const fs::path ROOT = PROJECT_ROOT;
const fs::path INPUTS = ROOT / "tests" / "inputs";
const fs::path EXPECTED = ROOT / "tests" / "expected";
const fs::path OUT = ROOT / "test_runs";
const fs::path WORK = OUT / "work";
const std::vector<std::string> SAVED_FILES = {"Grades.dat", "Grades.out", "Grades.trn", "screen.txt"};
const std::vector<std::string> CHECKED_FILES = {"Grades.dat", "Grades.out", "Grades.trn"};

// run the grade book once, reading input from inputFile, inside the WORK folder
void runGradeBook(const fs::path& inputFile){
    std::ifstream in(inputFile);
    std::ofstream screen(WORK / "screen.txt");
    std::streambuf* realCin = std::cin.rdbuf(in.rdbuf());
    std::streambuf* realCout = std::cout.rdbuf(screen.rdbuf());
    fs::path realFolder = fs::current_path();
    fs::current_path(WORK);  // Menu opens the Grades files in the current folder
    {
        Menu menu;
        menu.openMenu();
    }  // menu is destroyed here, which closes and saves Grades.out and Grades.trn
    fs::current_path(realFolder);
    std::cout.rdbuf(realCout);
    std::cin.rdbuf(realCin);
    std::cin.clear();  // the run may have left cin at end of file
}

// read a file's lines, or return false if the file doesn't exist
bool readLines(const fs::path& path, std::vector<std::string>& lines){
    std::ifstream in(path);
    if(!in){
        return false;
    }
    std::string line;
    while(std::getline(in, line)){
        // ignore Windows line endings, so files compare equal on any system
        if(!line.empty() && line.back() == '\r'){
            line.pop_back();
        }
        lines.push_back(line);
    }
    return true;
}

// compare one run's files with the expected copies; returns the problems found
std::vector<std::string> compareRun(const std::string& name){
    std::vector<std::string> problems;
    for(const std::string& fileName : CHECKED_FILES){
        std::vector<std::string> expected;
        std::vector<std::string> actual;
        bool haveExpected = readLines(EXPECTED / name / fileName, expected);
        bool haveActual = readLines(OUT / name / fileName, actual);
        if(!haveExpected || !haveActual){
            if(haveExpected != haveActual){
                problems.push_back(fileName + (haveExpected ? ": not written" : ": no expected copy"));
            }
            continue;
        }
        // report the first line that differs
        size_t lineCount = std::max(expected.size(), actual.size());
        for(size_t i = 0; i < lineCount; i++){
            std::string want = i < expected.size() ? expected[i] : "(end of file)";
            std::string got = i < actual.size() ? actual[i] : "(end of file)";
            if(want != got){
                problems.push_back(fileName + " line " + std::to_string(i + 1) + ":\n    expected: "
                                   + want + "\n    got:      " + got);
                break;
            }
        }
    }
    return problems;
}

int main(int argc, char* argv[]){
    bool update = argc > 1 && std::string(argv[1]) == "--update";

    // start from a clean folder, with no Grades.dat from earlier test runs.
    // Windows (or Google Drive syncing) can keep an empty folder locked, so
    // leftover empty folders are reused rather than treated as an error.
    std::error_code ignored;
    fs::remove_all(OUT, ignored);
    fs::create_directories(WORK);

    std::vector<fs::path> inputs;
    for(const fs::directory_entry& entry : fs::directory_iterator(INPUTS)){
        if(entry.path().extension() == ".txt"){
            inputs.push_back(entry.path());
        }
    }
    std::sort(inputs.begin(), inputs.end());

    bool allPassed = true;
    for(const fs::path& input : inputs){
        std::string name = input.stem().string();
        runGradeBook(input);

        // keep a copy of this run's files before the next run rewrites them
        fs::path runFolder = OUT / name;
        fs::create_directories(runFolder);
        for(const std::string& fileName : SAVED_FILES){
            if(fs::exists(WORK / fileName)){
                fs::copy_file(WORK / fileName, runFolder / fileName, fs::copy_options::overwrite_existing);
            }
        }

        std::string label = name + std::string(name.size() < 28 ? 28 - name.size() : 1, ' ');
        if(update){
            fs::create_directories(EXPECTED / name);
            for(const std::string& fileName : CHECKED_FILES){
                if(fs::exists(runFolder / fileName)){
                    fs::copy_file(runFolder / fileName, EXPECTED / name / fileName,
                                  fs::copy_options::overwrite_existing);
                }
            }
            std::cout << label << "expected output updated" << std::endl;
            continue;
        }

        std::vector<std::string> problems = compareRun(name);
        std::cout << label << (problems.empty() ? "PASS" : "FAIL") << std::endl;
        for(const std::string& problem : problems){
            std::cout << "  " << problem << std::endl;
        }
        if(!problems.empty()){
            allPassed = false;
        }
    }

    fs::remove(WORK / "screen.txt", ignored);
    std::cout << "Output files for each run are in test_runs/." << std::endl;
    return allPassed ? 0 : 1;
}

File: tests/run_tests.py
"""Test driver for the grade book program.

Runs the program once per file in tests/inputs (in name order), feeding the file as keyboard input. All runs share one working folder, so the Grades.dat saved by one run is loaded by the next, like running the program several times by hand. After each run, Grades.dat, Grades.out, Grades.trn and the screen output are copied to test_runs/<run name>/, since the program rewrites them every run.

Each run's Grades files are then compared with the known-good copies in tests/expected/<run name>/, and the run passes if they match.
After checking a deliberate change by hand, run with --update to save the new output as the expected output.

usage: python tests/run_tests.py [--update] [path/to/Program4]
"""
import shutil
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
INPUTS = ROOT / "tests" / "inputs"
EXPECTED = ROOT / "tests" / "expected"
OUT = ROOT / "test_runs"
WORK = OUT / "work"
TIME_LIMIT = 10  # seconds per run
SAVED_FILES = ["Grades.dat", "Grades.out", "Grades.trn", "screen.txt"]
CHECKED_FILES = ["Grades.dat", "Grades.out", "Grades.trn"]  # screen.txt is not compared
EXE_CANDIDATES = ["build/Debug/Program4.exe", "build/Program4.exe", "build/Program4"]


def find_program(args):
    if args:
        exe = Path(args[0]).resolve()
        return exe if exe.is_file() else None
    for candidate in EXE_CANDIDATES:
        exe = ROOT / candidate
        if exe.is_file():
            return exe
    return None


def read_lines(path):
    # text mode, so Windows and Unix line endings compare equal
    return path.read_text().splitlines() if path.exists() else None


def compare_run(name):
    """Return a list of problems, empty if this run matches the expected output."""
    problems = []
    for file_name in CHECKED_FILES:
        expected = read_lines(EXPECTED / name / file_name)
        actual = read_lines(OUT / name / file_name)
        if expected == actual:
            continue
        if expected is None or actual is None:
            problems.append(f"{file_name}: {'no expected copy' if expected is None else 'not written'}")
            continue
        # report the first line that differs
        for i in range(max(len(expected), len(actual))):
            want = expected[i] if i < len(expected) else "(end of file)"
            got = actual[i] if i < len(actual) else "(end of file)"
            if want != got:
                problems.append(f"{file_name} line {i + 1}:\n    expected: {want}\n    got:      {got}")
                break
    return problems


def main():
    args = sys.argv[1:]
    update = "--update" in args
    args = [arg for arg in args if arg != "--update"]

    exe = find_program(args)
    if exe is None:
        print("Program not found. Build it first with: cmake --build build")
        return 1

    shutil.rmtree(OUT, ignore_errors=True)
    if any(path.is_file() and path.name != "desktop.ini" for path in OUT.rglob("*")):
        print(f"Could not clear {OUT}. Close anything using it and try again.")
        return 1
    WORK.mkdir(parents=True, exist_ok=True)

    all_passed = True
    for input_file in sorted(INPUTS.glob("*.txt")):
        name = input_file.stem
        with open(input_file) as stdin, open(WORK / "screen.txt", "w") as stdout:
            try:
                subprocess.run([str(exe)], stdin=stdin, stdout=stdout, cwd=WORK,
                               timeout=TIME_LIMIT)
            except subprocess.TimeoutExpired:
                print(f"{name:<28} FAIL\n  did not finish in {TIME_LIMIT}s (does the input end with Q?)")
                all_passed = False

        # keep a copy of this run's files before the next run rewrites them
        run_dir = OUT / name
        run_dir.mkdir(exist_ok=True)
        for file_name in SAVED_FILES:
            if (WORK / file_name).exists():
                shutil.copy(WORK / file_name, run_dir)

        if update:
            (EXPECTED / name).mkdir(parents=True, exist_ok=True)
            for file_name in CHECKED_FILES:
                if (run_dir / file_name).exists():
                    shutil.copy(run_dir / file_name, EXPECTED / name)
            print(f"{name:<28} expected output updated")
            continue

        problems = compare_run(name)
        print(f"{name:<28} {'PASS' if not problems else 'FAIL'}")
        for problem in problems:
            print("  " + problem)
        if problems:
            all_passed = False

    (WORK / "screen.txt").unlink(missing_ok=True)
    print("Output files for each run are in test_runs/.")
    return 0 if all_passed else 1


if __name__ == "__main__":
    sys.exit(main())
